File: apps/swift-ios/Core/ProviderUsageLimits.swift
import Foundation

/// Optional enrichment of ServerProviderSnapshot; an older server omits it.
public struct ServerProviderUsageLimits: Codable, Equatable, Sendable {
    public struct Window: Codable, Equatable, Sendable, Identifiable {
        public let id: String
        public let kind: String
        public let label: String
        public let usedPercent: Double
        public let resetsAt: String?
        public let windowDurationMins: Int?
    }
    public struct Unavailable: Codable, Equatable, Sendable {
        public let reason: String
        public let message: String?
    }
    public struct ResetCredits: Codable, Equatable, Sendable {
        public let availableCount: Int
        public let nextExpiresAt: String?
    }
    public let resetCredits: ResetCredits?
    public let checkedAt: String
    public let windows: [Window]
    public let unavailable: Unavailable?
}

public struct ProviderConsumeResetCreditResult: Codable, Equatable, Sendable {
    public let outcome: String
    public let warning: String?

    public var message: String {
        if let warning { return warning }
        switch outcome {
        case "reset": return "Reset applied. Your windows have cleared."
        case "nothingToReset": return "Nothing to reset right now."
        case "noCredit": return "No reset credit left."
        case "alreadyRedeemed": return "That credit was already redeemed."
        default: return "The provider reported: \(outcome). Refresh to check your limits."
        }
    }
}

File: packages/contracts/src/providerUsageLimits.ts
import * as Schema from "effect/Schema";
import { ProviderInstanceId } from "./providerInstance.ts";
import {
  ForwardCompatibleArray,
  IsoDateTime,
  NonNegativeInt,
  TrimmedNonEmptyString,
} from "./baseSchemas.ts";

export const ServerProviderUsageWindow = Schema.Struct({
  id: TrimmedNonEmptyString,
  kind: Schema.Literals(["session", "weekly", "monthly", "other"]),
  label: TrimmedNonEmptyString,
  usedPercent: Schema.Number.check(Schema.isBetween({ minimum: 0, maximum: 100 })),
  resetsAt: Schema.optional(IsoDateTime),
  windowDurationMins: Schema.optional(NonNegativeInt),
});
export type ServerProviderUsageWindow = typeof ServerProviderUsageWindow.Type;

export const ServerProviderResetCredits = Schema.Struct({
  availableCount: NonNegativeInt,
  nextExpiresAt: Schema.optional(IsoDateTime),
});
export type ServerProviderResetCredits = typeof ServerProviderResetCredits.Type;

/** Optional snapshot enrichment. Failed probes never imply an unused allowance. */
export const ServerProviderUsageLimits = Schema.Struct({
  checkedAt: IsoDateTime,
  windows: ForwardCompatibleArray(ServerProviderUsageWindow),
  resetCredits: Schema.optional(ServerProviderResetCredits),
  unavailable: Schema.optional(
    Schema.Struct({
      reason: Schema.Literals(["unsupported", "probeFailed"]),
      message: Schema.optional(TrimmedNonEmptyString),
    }),
  ),
});
export type ServerProviderUsageLimits = typeof ServerProviderUsageLimits.Type;

export const ProviderConsumeResetCreditInput = Schema.Struct({ instanceId: ProviderInstanceId });
export type ProviderConsumeResetCreditInput = typeof ProviderConsumeResetCreditInput.Type;
export const ProviderConsumeResetCreditOutcome = Schema.Literals([
  "reset",
  "nothingToReset",
  "noCredit",
  "alreadyRedeemed",
]);
export type ProviderConsumeResetCreditOutcome = typeof ProviderConsumeResetCreditOutcome.Type;
export const ProviderConsumeResetCreditResult = Schema.Struct({
  outcome: ProviderConsumeResetCreditOutcome,
  warning: Schema.optional(TrimmedNonEmptyString),
});
export type ProviderConsumeResetCreditResult = typeof ProviderConsumeResetCreditResult.Type;

/** Account setup/action failure shared with newer provider clients. */
export class ProviderSetupError extends Schema.TaggedErrorClass<ProviderSetupError>()(
  "ProviderSetupError",
  {
    instanceId: ProviderInstanceId,
    operation: Schema.String,
    detail: TrimmedNonEmptyString,
    cause: Schema.optional(Schema.Defect()),
  },
) {
  override get message(): string {
    return this.detail;
  }
}
